rfcs/45-stageless.md

```diff
@@ -0,0 +1,584 @@
+6. Add atomic ordering constraints
+7. Add run criteria
+   1. Create `IntoRunCriteriaSystem` machinery
+   2. Store the run criteria of each `ConfiguredSystem` in the schedule
+   3. Add atomic ordering constraints
+   4. Check the values of the run-criteria of systems before deciding whether it can be run
+   5. Add methods to configure run criteria
+8. Add states
+   1. Create `State` trait
+   2. Implement on-update states as sugar for simple run criteria
+   3. Create on-enter and on-exit schedules
+   4. Create sugar for adding systems to these schedules
+9. Rename "system chaining" to "system handling"
+   1. Usage was very confusing for new users
+   2. Almost exclusively used for error handling
+   3. New concept of "systems with a linear graph of ordering constraints between them" is naturally described as a chain
+10. Add new examples
+    1. Complex control flow with supplementary schedules
+    2. Fixed time-step pattern
+11. Port the entire engine to the new paradigm
+    1. We almost certainly need to port the improved ambiguity checker over to make this reliable
+
+Given the massive scope, that sounds relatively straightforward!
+However, doing so will break approximately the entire engine, and tests will not pass again until step 10.
+
+Lets explore some of the more critical and challenging details.
+
+### Storing and configuring systems
+
+Schedules store systems in a configured, initialized form.
+
+```rust
+struct Schedule{
+   // Each schedule is associated with exactly one `World`, as systems must be initiliazed on it
+   world_id: WorldId
+   // We need to be able to quickly look up specific systems and iterate over them
+   // `SystemId` should be a unique opaque identifier, generated on system insertion
+   // We cannot use a Vec + a `usize` index as the identifier,
+   // as the identifier would not be stable when systems were removed, risking serious bugs
+   systems: HashMap<SystemId, ConfiguredSystem>,
+   // Label configuration must be stored at the schedule level,
+   // rather than attached to specific label instances
+   labels: HashMap<Box<dyn SystemLabel>, SystemConfig>,
+   // Used to quickly look up system ids by the type ids of their underlying function
+   // when checking atomic ordering constraints
+   system_function_type_map: HashMap<TypeId, Vec<SystemId>>,
+   // Used to check if mutation is allowed
+   currently_running: bool,
+}
+```
+
+Configured systems are composed of a raw system and its configuration.
+
+```rust
+struct ConfiguredSystem{
+   raw_system: RawSystem,
+   config: SystemConfig,
+   // All of the ordering constraints from `SystemConfig` 
+   // are converted into individual systems.
+   // Only the systems that this system are after are stored here.
+   // This is what is actually checked by the scheduler to see if this system can be run.
+   // All if-needed orderings are either converted to strict orderings at this step or removed.
+   cached_ordering: CachedOrdering, 

```

Comment:
This should probably be pulled out and stored densely on the schedule; `ConfiguredSystem` cannot be stack allocated.

Reply:
I'm not sure it's worth quibbling over how the data will be organized at this point. This and other data organization will need to be determined by whoever implements the new executor.